### PR TITLE
HTTP-63 Add caching

### PR DESCRIPTION
**Pull request by grzegorz8**
#### Description

Introduce caching similar to the Flink JDBC connector, so long running synchronous Flink jobs can cache content resulting in less API calls.

Resolves https://github.com/getindata/flink-http-connector/issues/63

I took over this PR: https://github.com/getindata/flink-http-connector/pull/94 due to author inactivity.

##### PR Checklist
- [x] Tests added
- [x] [Changelog](CHANGELOG.md) updated

### CODE REVIEW

File: src/main/java/com/getindata/connectors/http/internal/table/lookup/AsyncHttpTableLookupFunction.java [hunk 1]
```diff
@@ -73,29 +71,27 @@ public void open(FunctionContext context) throws Exception {
             );
     }
 
-    public void eval(CompletableFuture<Collection<RowData>> resultFuture, Object... keys) {
-
-        CompletableFuture<Optional<RowData>> future = new CompletableFuture<>();
-        future.completeAsync(() -> decorate.lookupByKeys(keys), pullingThreadPool);
+    @Override
+    public CompletableFuture<Collection<RowData>> asyncLookup(RowData keyRow) {
```
Comment (grzegorz8) [inline src/main/java/com/getindata/connectors/http/internal/table/lookup/AsyncHttpTableLookupFunction.java:75]: @kristoffSC Please note that `AsyncLookupFunction#asyncLookup(RowData)` is called by `org.apache.flink.table.functions.AsyncLookupFunction#eval` which has the following implementation:

```java
    public final void eval(CompletableFuture<Collection<RowData>> future, Object... keys) {
        GenericRowData keyRow = GenericRowData.of(keys);
        asyncLookup(keyRow)
                .whenComplete(
                        (result, exception) -> {
                            if (exception != null) {
                                future.completeExceptionally(
                                        new TableException(
                                                String.format(
                                                        "Failed to asynchronously lookup entries with key '%s'",
                                                        keyRow),
                                                exception));
                                return;
                            }
                            future.complete(result);
                        });
    }
```

It is very similar to what we have in `asyncLookup` but it uses default thread pool while here we have `publishingThreadPool`.

Comment (kristoffSC) [reply src/main/java/com/getindata/connectors/http/internal/table/lookup/AsyncHttpTableLookupFunction.java:75]: I dont think it uses default pool. From what I see it depends on the impl of `asyncLookup` so it is up to the implmenenbtation. The flink's `CachingAsyncLookupFunction` also delegated to `delegate` impl. The `whenComplete` will be executed on whatever thread was used to call `asyncLookup`

**Review by kristoffSC — APPROVED** [review]
Im super happy to see this finally added to the connector :)
Thanks @grzegorz8 !!!

File: src/main/java/com/getindata/connectors/http/internal/table/lookup/HttpLookupTableSource.java [hunk 2]
```diff
@@ -46,17 +51,20 @@ public class HttpLookupTableSource
...
+    private final LookupCache cache;
 
     public HttpLookupTableSource(
             DataType physicalRowDataType,
             HttpLookupConfig lookupConfig,
             DecodingFormat<DeserializationSchema<RowData>> decodingFormat,
-            DynamicTableFactory.Context dynamicTablecontext) {
-
+            DynamicTableFactory.Context dynamicTablecontext,
+            @Nullable LookupCache cache) {
```
Comment (kristoffSC) [inline src/main/java/com/getindata/connectors/http/internal/table/lookup/HttpLookupTableSource.java:62]: nit:
Im still not super happy that we can pass null here like described in @davidradl [PR](https://github.com/getindata/flink-http-connector/pull/94#discussion_r1636997742) but maybe its just me trying to over complicate things.

File: src/main/java/com/getindata/connectors/http/internal/table/lookup/AsyncHttpTableLookupFunction.java (hunk 1 above)
Comment (grzegorz8) [reply src/main/java/com/getindata/connectors/http/internal/table/lookup/AsyncHttpTableLookupFunction.java:75]: You are right, thanks for double-checking it.

File: src/main/java/com/getindata/connectors/http/internal/table/lookup/HttpLookupTableSource.java (hunk 2 above)
Comment (grzegorz8) [reply src/main/java/com/getindata/connectors/http/internal/table/lookup/HttpLookupTableSource.java:62]: The same approach is used in flink-connector-jdbc: https://github.com/apache/flink-connector-jdbc/blob/v3.2/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java#L85

**Comment by grzegorz8** [comment]
@kristoffSC Build for 1.18.1 fails due to:

```

Error: ENOENT: no such file or directory, open '/home/runner/work/flink-http-connector/flink-http-connector/target/site/jacoco/jacoco.xml'
    at async open (node:internal/fs/promises:636:25)
    at async Object.readFile (node:internal/fs/promises:1246:14)
    at async /home/runner/work/_actions/madrapps/jacoco-report/v1.2/dist/index.js:12458:25
    at async Promise.all (index 0) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/home/runner/work/flink-http-connector/flink-http-connector/target/site/jacoco/jacoco.xml'
}
```
Any thoughts?

**Review by MarekMaj — APPROVED** [review]

**Comment by MarekMaj** [comment]
> @kristoffSC Build for 1.18.1 fails due to:
> 
> ```
> 
> Error: ENOENT: no such file or directory, open '/home/runner/work/flink-http-connector/flink-http-connector/target/site/jacoco/jacoco.xml'
>     at async open (node:internal/fs/promises:636:25)
>     at async Object.readFile (node:internal/fs/promises:1246:14)
>     at async /home/runner/work/_actions/madrapps/jacoco-report/v1.2/dist/index.js:12458:25
>     at async Promise.all (index 0) {
>   errno: -2,
>   code: 'ENOENT',
>   syscall: 'open',
>   path: '/home/runner/work/flink-http-connector/flink-http-connector/target/site/jacoco/jacoco.xml'
> }
> ```
> 
> Any thoughts?

Looks like jacoco report is not generated, logs show:
```
05:46:48,640 [INFO] --- jacoco-maven-plugin:0.8.7:report (jacoco-report) @ flink-http-connector ---
05:46:48,643 [INFO] Skipping JaCoCo execution due to missing execution data file.
```

The reason may be related to this added configuration of maven-surefire-plugin
```
<argLine>--add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED</argLine>
```

[Documentation](https://www.eclemma.org/jacoco/trunk/doc/prepare-agent-mojo.html) suggest to add `@{argLine}` in maven-surefire-plugin if we want to keep this arguments

**Comment by grzegorz8** [comment]
> > @kristoffSC Build for 1.18.1 fails due to:
> > ```
> > 
> > Error: ENOENT: no such file or directory, open '/home/runner/work/flink-http-connector/flink-http-connector/target/site/jacoco/jacoco.xml'
> >     at async open (node:internal/fs/promises:636:25)
> >     at async Object.readFile (node:internal/fs/promises:1246:14)
> >     at async /home/runner/work/_actions/madrapps/jacoco-report/v1.2/dist/index.js:12458:25
> >     at async Promise.all (index 0) {
> >   errno: -2,
> >   code: 'ENOENT',
> >   syscall: 'open',
> >   path: '/home/runner/work/flink-http-connector/flink-http-connector/target/site/jacoco/jacoco.xml'
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any thoughts?
> 
> Looks like jacoco report is not generated, logs show:
> 
> ```
> 05:46:48,640 [INFO] --- jacoco-maven-plugin:0.8.7:report (jacoco-report) @ flink-http-connector ---
> 05:46:48,643 [INFO] Skipping JaCoCo execution due to missing execution data file.
> ```
> 
> The reason may be related to this added configuration of maven-surefire-plugin
> 
> ```
> <argLine>--add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED</argLine>
> ```
> 
> [Documentation](https://www.eclemma.org/jacoco/trunk/doc/prepare-agent-mojo.html) suggest to add `@{argLine}` in maven-surefire-plugin if we want to keep this arguments

Thanks. It helped.

**Pull request merged**